miqi/agent/tools/graph_render.py

```
@@ -0,0 +1,814 @@
+            if ref is not None and str(ref) not in ids:
+                missing.append(f"边 {e.get('from')}→{e.get('to')} 引用了不存在的节点 {ref}")
+    return missing
+
+
+# ── 分层布局（DAG 通用：最长路径分层 + 层内垂直堆叠）────────────────────
+def _compute_layers(graph: dict[str, Any]) -> tuple[list[str], dict[str, int]]:
+    """返回 (主链拓扑序, id→层号)。implicit/failure 节点放最后。"""
+    nodes = graph["nodes"]
+    ids = _node_ids(graph)
+
+    # 隐式汇聚节点（failure/implicit）不参与主链分层，置于末尾
```

Ruff (0.16.1)

[warning] Comment contains ambiguous `（` (FULLWIDTH LEFT PARENTHESIS). Did you mean `(` (LEFT PARENTHESIS)?

(RUF003)